lib/Transforms/MemoryBanking.cpp

```diff
@@ -409,6 +409,151 @@ struct BankAffineStorePattern
+                                unsigned bankingDimension) const {
+    int64_t rank = shape.size();
+    // Initialize SubView offsets, sizes, strides for each dimension with
+    // default values.
+    SmallVector<OpFoldResult> offsets(rank, rewriter.getIndexAttr(0));
+    SmallVector<OpFoldResult> sizes(rank);
+    SmallVector<OpFoldResult> strides(rank, rewriter.getIndexAttr(1));
+
+    for (int64_t d = 0; d < rank; ++d) {
+      int64_t dimSize = shape[d];
+      if (d == bankingDimension) {
+        // Round-robin partition: offset = bankIndex, size = chunkSize, stride =
+        // bankingFactor
+        int64_t chunkSize = dimSize / (int64_t)bankingFactor;
+        offsets[d] = rewriter.getIndexAttr(bankIndex);
+        sizes[d] = rewriter.getIndexAttr(chunkSize);
+        strides[d] = rewriter.getIndexAttr(bankingFactor);
+      } else {
+        // Non-banked dimension: copy the entire range, unit stride
+        offsets[d] = rewriter.getIndexAttr(0);
+        sizes[d] = rewriter.getIndexAttr(dimSize);
+        strides[d] = rewriter.getIndexAttr(1);
+      }
+    }
+
+    return rewriter.create<memref::SubViewOp>(loc, mem, offsets, sizes,
+                                              strides);
+  }
+
+  /// Copy from source memory reference to destination memory reference when the
+  /// source memory is banked into `srcBanks` and `dstMem` is unbanked. All
+  /// `srcBanks` have the same MemRefType; the unbanked `dstMem` has the same
+  /// shape as the source's "original" shape.
+  LogicalResult copyFromBanksToUnbanked(PatternRewriter &rewriter, Location loc,
+                                        ArrayRef<Value> srcBanks, Value dstMem,
+                                        uint64_t bankingFactor,
+                                        unsigned bankingDimension) const {
+    MemRefType dstType = dyn_cast<MemRefType>(dstMem.getType());
+    if (!dstType || dstType.getRank() <= bankingDimension)
+      return failure();
+
+    ArrayRef<int64_t> dstShape = dstType.getShape();
+    int64_t chunkSize = dstShape[bankingDimension] / (int64_t)bankingFactor;
+    if (chunkSize <= 0)
+      return failure();
+
+    // For each bank i in [0..bankingFactor), create a subview of the source
+    //    that picks out the "i-th round-robin slice" along the banking
+    //    dimension.
+    for (size_t i = 0; i < bankingFactor; ++i) {
+      auto subView = createRoundRobinSubview(rewriter, loc, dstMem, dstShape, i,
+                                             bankingFactor, bankingDimension);
+      rewriter.create<memref::CopyOp>(loc, srcBanks[i], subView);
+    }
+
+    return success();
+  }
+
+  /// Copy from unbanked source memref `srcMem` to destination memory references
+  /// `dstBanks`, where each bank has the same MemRefType, and `srcMem` has the
+  /// shape of the "original" unbanked array.
+  LogicalResult copyFromUnbankedToBanks(PatternRewriter &rewriter, Location loc,
```

Comment:
This seems to be almost exactly the same as the function above. Can we add additional arguments or template parameters to avoid having two nearly duplicate functions? WDYT?